packages/flutter/lib/src/gestures/selection_recognizers.dart

```diff
@@ -0,0 +1,714 @@
+      } else if (_dragState == _GestureState.accepted) {
+        _giveUpPointer(event.pointer);
+      }
+    } else if (event is PointerCancelEvent){
+      _giveUpPointer(event.pointer);
+    }
+  }
+
+  @override
+  void rejectGesture(int pointer) {
+    if (pointer != primaryPointer) {
+      return;
+    }
+
+    _stopDeadlineTimer();
+    _giveUpPointer(pointer);
+
+    // Reset down and up when the recognizer has been rejected.
+    // This prevents an erroneous _up being sent when this recognizer is
+    // accepted for a drag, following a previous rejection.
+    _resetTaps();
+    consecutiveTapReset();
+    _initialButtons = null;
+  }
+
+  @override
+  void dispose() {
+    _stopDeadlineTimer();
+    consecutiveTapReset();
+    super.dispose();
+  }
+
+  @override
+  String get debugDescription => 'tap_and_drag';
+
+  void _acceptDrag(PointerMoveEvent event) {
+    _checkTapCancel();
+    if (dragStartBehavior == DragStartBehavior.start) {
+      _initialPosition = _initialPosition + OffsetPair(global: event.delta, local: event.localDelta);
+    }
+    _checkStart(event);
+    if (event.localDelta != Offset.zero) {
+      final Matrix4? localToGlobal = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
+      final Offset correctedLocalPosition = _initialPosition.local + event.localDelta;
+      final Offset globalUpdateDelta = PointerEvent.transformDeltaViaPositions(
+        untransformedEndPosition: correctedLocalPosition,
+        untransformedDelta: event.localDelta,
+        transform: localToGlobal,
+      );
+      final OffsetPair updateDelta = OffsetPair(local: event.localDelta, global: globalUpdateDelta);
+      _correctedPosition = _initialPosition + updateDelta; // Only adds delta for down behaviour
+      _checkUpdate(event);
+      _correctedPosition = null;
+    }
+  }
+
+  void _checkDrag(PointerMoveEvent event) {
+    final Matrix4? localToGlobalTransform = event.transform == null ? null : Matrix4.tryInvert(event.transform!);
+    _globalDistanceMoved += PointerEvent.transformDeltaViaPositions(
+      transform: localToGlobalTransform,
+      untransformedDelta: event.localDelta,
+      untransformedEndPosition: event.localPosition
+    ).distance * 1.sign;
+    if (_hasSufficientGlobalDistanceToAccept(event.kind, gestureSettings?.touchSlop)) {
+      if (event.buttons == kSecondaryButton) {
+        // Reject a right click drag.
+        resolve(GestureDisposition.rejected);
+        return;
+      }
+      _start = event;
+      _dragState = _GestureState.accepted;
+      resolve(GestureDisposition.accepted);
+    }
+  }
+
+  void _checkTapDown(PointerDownEvent event) {
+    if (_sentTapDown) {
+      return;
+    }
+
+    final TapDownDetails details = TapDownDetails(
+      globalPosition: event.position,
+      localPosition: event.localPosition,
+      kind: getKindForPointer(event.pointer),
+    );
+
+    incrementConsecutiveTapCountOnDown(details.globalPosition);
```

Comment:
I imagine the _ConsecutiveTapMixin to be more self contain. can the _ConsecutiveTapMixin to be mix on OneSequenceGestureRecognizer so that it can override addAllowGesture, accept gesture and the rejectGesture to update its own state? TapAndDragGestureRecognizer can just query the consecutiveTapCount when the drag starts

Reply:
I think we have to continue to query the consecutiveTapCount in the onTapDown. The primary reason being there can be a double click hold and then drag. If we query during drag start the consecutiveTapCount may have been reset by the `Timer` timeout callback.

Reply:
you can override handleEvent to get record most recent down event?